src/NuGet.Clients/NuGet.VisualStudio.Internal.Contracts/Formatters/NuGetVersionFormatter.cs

```diff
@@ -0,0 +1,96 @@
+
+        private NuGetVersionFormatter()
+        {
+        }
+
+        public NuGetVersion? Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
+        {
+            if (reader.TryReadNil())
+            {
+                return null;
+            }
+
+            // stack overflow mitigation - see https://github.com/neuecc/MessagePack-CSharp/security/advisories/GHSA-7q36-4xx7-xcxf
+            options.Security.DepthStep(ref reader);
+
+            try
+            {
+                int propertyCount = reader.ReadMapHeader();
+
+                string? versionString = null;
+                string? metadata = null;
+                string? originalVersion = null;
+                IEnumerable<string>? releaseLabels = null;
+
+                for (int propertyIndex = 0; propertyIndex < propertyCount; propertyIndex++)
+                {
+                    switch (reader.ReadString())
+                    {
+                        case VersionPropertyName:
+                            versionString = reader.ReadString();
+                            break;
+                        case MetadataPropertyName:
+                            metadata = reader.ReadString();
+                            break;
+                        case ReleaseLabelsPropertyName:
+                            releaseLabels = options.Resolver.GetFormatter<IEnumerable<string>>().Deserialize(ref reader, options);
+                            break;
+                        case OriginalVersionPropertyName:
+                            originalVersion = reader.ReadString();
+                            break;
+                        default:
+                            reader.Skip();
+                            break;
+                    }
+                }
+
+                return new NuGetVersion(new Version(versionString), releaseLabels, metadata, originalVersion);
+            }
+            finally
+            {
+                // stack overflow mitigation - see https://github.com/neuecc/MessagePack-CSharp/security/advisories/GHSA-7q36-4xx7-xcxf
+                reader.Depth--;
+            }
+        }
+
+        public void Serialize(ref MessagePackWriter writer, NuGetVersion? value, MessagePackSerializerOptions options)
+        {
+            if (value == null)
+            {
+                writer.WriteNil();
+                return;
+            }
+
+            writer.WriteMapHeader(4);
+            writer.Write(VersionPropertyName);
+            writer.Write(value.Version.ToString());
+            writer.Write(MetadataPropertyName);
+            writer.Write(value.Metadata);
+            writer.Write(OriginalVersionPropertyName);
+            writer.Write(value.OriginalVersion);
+            writer.Write(ReleaseLabelsPropertyName);
+            options.Resolver.GetFormatter<IEnumerable<string>>().Serialize(ref writer, value.ReleaseLabels, options);
```

Comment:
Does this have an advantage over using only `value.OriginalVersion` and `NuGetVersion.Parse`?

Reply:
Not that I can see, changed it to use that pattern.